scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,211 @@
 package org.scalafmt.sbt
+
+import org.scalafmt.config.{Config, ScalafmtConfig}
+import org.scalafmt.{Formatted, Scalafmt}
+import sbt.Keys._
+import sbt.{Def, _}
+import complete.DefaultParsers._
+import metaconfig.Configured
+import sbt.util.Logger
+
 import scala.meta.internal.tokenizers.PlatformTokenizerCache
-import sbt._
-import Keys._
+import scala.util.{Failure, Success, Try}
 
 object ScalafmtPlugin extends AutoPlugin {
   override def trigger: PluginTrigger = allRequirements
+
   object autoImport {
-    val scalafmt: Command =
-      Command.args("scalafmt", "run the scalafmt command line interface.") {
-        case (state, args) =>
-          org.scalafmt.cli.Cli.main("--non-interactive" +: args.toArray)
-          PlatformTokenizerCache.megaCache.clear()
-          state
+    val scalafmt = taskKey[Unit]("Format Scala sources.")
+    val scalafmtCheck =
+      taskKey[Boolean](
+        "Fails if a Scala source is mis-formatted. Does not write to files.")
+    val scalafmtOnCompile =
+      settingKey[Boolean](
+        "Format Scala source files on compile, off by default. Warning: It formats all project sources.")
+    val scalafmtConfig = taskKey[Option[File]](
+      "Optional location of .scalafmt.conf file. If None the default config is used.")
+    val scalafmtSbt = taskKey[Unit](
+      "Format *.sbt and project/*.scala files for this sbt build.")
+    val scalafmtSbtCheck =
+      taskKey[Boolean](
+        "Fails if a *.sbt or project/*.scala source is mis-formatted. Does not write to files.")
+    val scalafmtOnly = inputKey[Unit]("Format only given file.")
+  }
+  import autoImport._
+
+  private val scalafmtDoFormatOnCompile =
+    taskKey[Unit]("Format Scala source files if scalafmtOnCompile is on.")
+
+  private val scalaConfig =
+    scalafmtConfig.map(_.map(Config.fromHoconFile(_) match {
```

Comment:
Could we at least cache this by the timestamp of the file? I feel it's unnecessary to reparse the config for every project/config on every compile.

Reply:
Here's what the IntelliJ plugin uses https://github.com/scalameta/scalafmt/blob/0f391aa351c59432d1cf6693b5658ab81d4de517/scalafmt-cli/src/main/scala/org/scalafmt/cli/StyleCache.scala

Reply:
I see, sure we can leave it for a separate PR :)

Reply:
>Here's what the IntelliJ plugin uses 

@olafurpg maybe simply reuse that object? It does just what we need

Reply:
Go ahead, we might want to move it to the `core` module